### ISSUE TITLE
RedundantModifierCheck: New property 'jdkVersion' to violate strictfp as redundant modifier if version LTE 17

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#10798]
http://checkstyle.sourceforge.net/config_modifier.html#RedundantModifier
Reason: since java 17 there is no need for this modifier, see details in [JEP 306](https://openjdk.java.net/jeps/306)

Suggestion is to add new property `jdkVersion` where user can define jdk version to let us make decision on dedundancy of `strictfp`. Before 17th not redundant, on 17 and afrter - redundant. No jdk specified, no violation on  `strictfp`.

```bash
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="error"/>

    <module name="TreeWalker">
        <module name="RedundantModifier">
         <property name="jdkVersion" value="17">
        </module>
    </module>
</module>

/var/tmp $ cat Test.java
class Test {
  void strictfp method(){} // violation
}
```
Expected - violation reported on line with violation comment

jdkversion values should be integers "1,2,3,4,5,6,7,8,9,10....", as well as support old maven versioning of 1.8, etc...

**Comment by AayushSaini101** [comment]
@strkkk  Sir i have made the all the required changes but I am facing an issue. can you please check
`[ERROR] Errors: 
[ERROR]   XdocsJavaDocsTest.testAllCheckSectionJavaDocs:182->examineCheckSection:209 »  ...
[INFO]
[ERROR] Tests run: 3631, Failures: 0, Errors: 1, Skipped: 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:49 min
[INFO] Finished at: 2021-09-15T13:44:21+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project checkstyle: There are test failures.
[ERROR]
[ERROR] Please refer to C:\Users\Aayush\Desktop\New folder (2)\checkstyle\target\surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
`

**Comment by strkkk** [comment]
@AayushSaini101 you need to correct javadoc and xdocs. You can run this test locally and see diff

**Comment by AayushSaini101** [comment]
Sir how to run these test cases locally .

**Comment by strkkk** [comment]
@AayushSaini101 open test class, navigate to failed test method and follow instructions https://www.jetbrains.com/help/idea/performing-tests.html

**Comment by AkMo3** [comment]
@strkkk After JDK 17 it is redundant, but it is still not redundant for earlier versions I believe. Shouldn't we add a check for the Java version and then change the acceptable token according to it?

**Comment by strkkk** [comment]
> 
> 
> @strkkk After JDK 17 it is redundant, but it is still not redundant for earlier versions I believe. Shouldn't we add a check for the Java version and then change the acceptable token according to it?

I think it is up to user to decide what java version they are using and do they need it or not. So I suggest to do not try to make check too smart and complicated.

There can be a different option - add some property that can provide java version for checkstyle, in this way we can make some adjustments for default configuration for checks.

**Comment by AkMo3** [comment]
@strkkk I was working on it and found that just changing the acceptable tokens wouldn't work as there are separate tests defined for separate tokens like `INTERFACE_DEF` and `ENUM_DEF`. Also, here `strictfp` is a modifier, not a definition. So rather than changing the acceptable token, I created the following function and checks all tokens

```java
/**
 * Checks if the ast has strictfp modifier in java environment with a version greater than 16.
 *
 * @param ast token AST.
 */
private void checkStrictfp(DetailAST ast) {
    final DetailAST methodModifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
    final double systemJavaVer = Double.parseDouble(System.getProperty("java.version")
            .substring(0, 2));
    final int strictFpRedundantJavaVer = 17;
    if (methodModifiers != null) {
        final DetailAST strictFpToken = methodModifiers.findFirstToken(TokenTypes.STRICTFP);
        if (systemJavaVer >= strictFpRedundantJavaVer && strictFpToken != null) {
            log(strictFpToken, MSG_KEY, strictFpToken.getText());
        }
    }
}
```

The issue seems to be resolved by this, what do you think?

**Comment by strkkk** [comment]
> Also, here `strictfp` is a modifier, not a definition

I really do not understand a desire to make things complicated. The fix for this issue is like this
1. Add token to list of acceptable token
2. add one if to visitToken method, see below
```
@Override
    public void visitToken(DetailAST ast) {
        if (ast.getType() == TokenTypes.STRICTFP) {
            // report violation
        }
       // rest of method
}
```
>  I created the following function and checks all tokens

Your approach has signficant flow - runtime version of java, that executes checkstyle code != source version of the code that is under inspection. For example, there is some common CI pipeline used by multiple projects, and its runtime is java 17. But code under check can be java 8.
Therefore IMO there is only one approach- let users configure check themselves.

**Comment by nmancus1** [comment]
> add some property that can provide java version for checkstyle

I think if we do this, we should open a new issue and discuss there. 

>The fix for this issue is like this

>1. Add token to list of acceptable tokens
>2. add one if to visitToken method, see below...

I agree with this, and I think we should add:

3. Add to check documentation a link to https://openjdk.java.net/jeps/306 with note that this is why `STRICTFP` has been added to the list of acceptable tokens, but not default. 


This will help users make an informed decision about inclusion of this token in their config, and eliminate the need for additional check properties such as `strictfpRedundant`.

**Comment by AkMo3** [comment]
@nmancus1 Changes are done, please take a glance

**Comment by pbludov** [comment]
Such a property should not be added per check, but one at the TreeWalker level.
I agree with @strkkk that this will be a difficult and confusing way.
We need to update all docs to show differences in default property values depending on the Java source level.

**Comment by romani** [comment]
if user is alread in jdk17, he can forbit usage of token by https://checkstyle.org/config_coding.html#IllegalToken
so easy.
```
    <module name="IllegalToken">
         <property name="tokens" value="STRICTFP"/>
    </module>
```
instead of (that do the same goal, violate modifier irrespective of jdk):
```
    <module name="RedundantModifier">
         <property name="tokens" value="STRICTFP"/>
    </module>
```

In this case it would be better to mention that RedundantModifier documentation that `strictfp` is not covered at all and IllegalToken should be used.

Execution of this Check on code base before jdk17 will create a lot of confusions. Something should know about target jdk to make decision on what is redundant what what is not.

**Comment by nmancus1** [comment]
> Execution of this Check on code base before jdk17 will create a lot of confusions.

Implementation at https://github.com/checkstyle/checkstyle/pull/10804 requires that user add `STRICTFP` token to their config manually, it is not default.

> if user is alread in jdk17, he can forbit usage of token by...

Yes, this is true. But, this requires more effort from the user, if they want violation to be informative to others (i.e. changing the error message).

Default message from `IllegalTokenCheck`:

(1) `Using STRICTFP is not allowed.`

vs.

Default message from `RedundantModifierCheck`:

(2) `Redundant STRICTFP modifier.`

IMO, 2 is more informative. If we do end up just adding a note in `RedundantModifierCheck` to use `IllegalTokenCheck`, we should also supply an example config in `IllegalTokenCheck` docs that gives a more descriptive error message for `STRICTFP` modifier.

**Comment by romani** [comment]
User can override message to make it same in wording

**Comment by strkkk** [comment]
@romani strictfp is a modifier (see [jls](https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.4)), so RedundantModifierCheck seems a perfect place for it. 

And yes, as mentioned above, STRICTFP token is not in list of default, to allow user to config it by themselves.

**Comment by romani** [comment]
It is good Check to do this - no disagreement.
We never used tokens for such behavior and this weird for me. It is more logical to make "jdkVersion" field, it has a lot of potential.
as it will be continue to be issue from jdk to jdk.

https://github.com/checkstyle/checkstyle/pull/10804#discussion_r757823390

**Comment by nmancus1** [comment]
> It is more logical to make "jdkVersion" field, it has a lot of potential

As TreeWalker property,
as @pbludov mentioned [above](https://github.com/checkstyle/checkstyle/issues/10798#issuecomment-922429238) ? Or just to this check?

**Comment by rnveach** [comment]
TreeWalker is not responsible for printing violations of JDK version. Its purpose is to parse and pass the parsing results on. It can print violations for parsing errors or children errors, but I don't see it being responsible for printing violations on redundant modifiers.

**Comment by nmancus1** [comment]
> TreeWalker is not responsible for printing violations of JDK version.

I think the idea was to have TreeWalker property like:

```xml

<module name="Translation">
  <property name="jdkVersion" value="17"/>
</module>

```
so that during check module initialization, TreeWalker can configure given check with this property. 

We could even extend this property to be passed to `JavaParser` by `TreeWalker` as well. This would make it possible to use predicates to "turn off" parts of our grammar that are not needed by a user. This idea was originally proposed by @pbludov during GSOC 2020.

**Comment by romani** [comment]
Jdk version in TreeWalker is too much. If we make parser be able to disable some features, it will force us to test a lot with different versions of jdk. If AST be different for each version, we will have to support all and it is over commitment to support.

This Check is changing each time java get big update. There was  a lot to change during jdk8. Probably something was changed due to others jdk, but all of them were compatible. Only now we have non compatible Check behavior.
I still believe that this Check will benefit from jdkVersion property.

**Comment by mahfouz72** [comment]
> almost whole work is done at [#10804](https://github.com/checkstyle/checkstyle/pull/10804), just need new property introduction.

I didn't read all these comments yet. but I assume we will have to make some property like this  ? 

> As a result of the aforementioned difference between Java 5 and Java 6, a property called javaFiveCompatibility is available. This property will only check classes, interfaces, etc. that do not contain the extends or implements keyword or are not anonymous classes. This means it only checks methods overridden from java.lang.Object. Java 5 Compatibility mode severely limits this check. It is recommended to only use it on Java 5 source.

from https://checkstyle.org/checks/annotation/missingoverride.html

**Comment by nrmancuso** [comment]
This check is becoming highly dependent on Java version. We must introduce a property like `jdkVersion` at this point.

**Comment by mahfouz72** [comment]
@romani @nrmancuso @rnveach please update the issue title to reflect the new change, we are not adding it as a token. we introduced a new property

**Comment by nrmancuso** [comment]
Not sure how I closed this, but it was a mistake for sure :)
